Lib/asyncio/unix_events.py

```diff
@@ -1266,7 +1266,13 @@ def is_active(self):
         return True
 
     def close(self):
-        pass
+        self._join_threads()
+
+    def _join_threads(self):
+        threads = [thread for thread in list(self._threads.values())
+                            if thread.is_alive()]
```

Comment:
As a minor optimization, would be worthwhile to change this to a generator expression? 
```
        threads = (thread for thread in list(self._threads.values())
                            if thread.is_alive())
```
There's no need to use a list here. The same applies to the `__del__()` method on line 1284.

Reply:
Adding to this, we could also change the list conversion to an iterable conversion. There's no reason to use a list since we're just iterating over the threads in the values view without adding or removing items.
```
         threads = (thread for thread in iter(self._threads.values())
                            if thread.is_alive())
```

Reply:
FWIW I think it would be acceptable to add a *private* methof to ThreadedChildWatcher `_join_threads()` and use it from our unit tests. No need to document it beyond just writing a doc string.  We can add a public API for this sometime later.

Reply:
>  We can add a public API for this sometime later.

@1st1  Yeah those was my thoughts as well. I think it made the most sense to include it as part of `close()`, since that's intended to be used as the Watcher cleanup method and fits well with the existing tests. I generally prefer to avoid modifying the regression tests as much as possible if there's not a functional benefit for doing so. 

Is there anything from the current state of the PR you'd like to see changed other than writing a docstring?

Reply:
@1st1 
> FWIW I think it would be acceptable to add a private methof to ThreadedChildWatcher _join_threads() and use it from our unit tests

If we were to call it explicitly from the unit test, how would you recommend doing so? The current `tearDown()` method is generically structured for all of the `ChildWatcher` classes:
```
        def tearDown(self):
            super().tearDown()
            policy = asyncio.get_event_loop_policy()
            watcher = policy.get_child_watcher()
            policy.set_child_watcher(None)
            watcher.attach_loop(None)
            watcher.close()
```

Would it work if we overrode the `tearDown()` method for the `ThreadedChildWatcher` test class?:

Current:
```

    class SubprocessThreadedWatcherTests(SubprocessWatcherMixin,
                                         test_utils.TestCase):

        Watcher = unix_events.ThreadedChildWatcher
```
Suggested:
```
    class SubprocessThreadedWatcherTests(SubprocessWatcherMixin,
                                         test_utils.TestCase):

        Watcher = unix_events.ThreadedChildWatcher

        def tearDown(self):
            super().tearDown()
            Watcher._join_threads()
```

I'm not sure if `_join_threads()` should be called earlier or if it's okay to just call it after SubprocessWatcherMixin.tearDown() is finished (as the above example does).